pkgs/_macro_host/lib/macro_host.dart

```diff
@@ -2,26 +2,91 @@
+///
+/// Tools that want to support macros, such as the Analyzer and the CFE, can
+/// do so by running a `MacroHost` and providing their own `MacroService`.
+class MacroHost implements MacroService {
   final MacroServer macroServer;
+  final ListOfServices services;
   final MacroBuilder macroBuilder = MacroBuilder();
   final MacroRunner macroRunner = MacroRunner();
-  late final HostEndpoint hostEndpoint;
 
-  MacroHost._(this.macroServer, this.hostEndpoint);
+  // TODO(davidmorgan): this should be per macro, as part of tracking per-macro
+  // lifecycle state.
+  Set<int>? _macroPhases;
+
+  MacroHost._(this.macroServer, this.services) {
+    services.services.insert(0, this);
+  }
+
+  /// Starts a macro host serving the provided [service].
+  ///
+  /// The service passed in should handle introspection RPCs, it does not need
+  /// to handle others.
+  ///
+  /// TODO(davidmorgan): make this split clearer, it should be in the protocol
+  /// definition somewhere which requests the host handles.
+  static Future<MacroHost> serve({required MacroService service}) async {
+    final listOfServices = ListOfServices();
+    listOfServices.services.add(service);
+    final server = await MacroServer.serve(service: listOfServices);
+    return MacroHost._(server, listOfServices);
+  }
+
+  /// Whether [name] is a macro according to that package's `pubspec.yaml`.
+  bool isMacro(File packageConfig, QualifiedName name) {
+    // TODO(language/3728): this is a placeholder, use package config when
+    // available.
+    return true;
+  }
 
-  static Future<MacroHost> serve({required MacroHostService service}) async {
-    final server = MacroServer(service: service);
-    final endpoint = await server.serve();
-    return MacroHost._(server, endpoint);
+  /// Determines which phases the macro implemented at [name] runs in.
+  Future<Set<int>> queryMacroPhases(
+      File packageConfig, QualifiedName name) async {
+    if (_macroPhases != null) return _macroPhases!;
+    final macroBundle = await macroBuilder.build(packageConfig, [name]);
+    macroRunner.run(macroBundle: macroBundle, endpoint: macroServer.endpoint);
+    // TODO(davidmorgan): wait explicitly for the MacroStartedRequest to
+    // arrive, remove this hard-coded wait.
+    await Future<void>.delayed(const Duration(seconds: 2));
+    return _macroPhases!;
   }
 
-  // TODO(davidmorgan): methods for integration with analyzer+CFE go here:
-  // check if an annotation is linked to a macro, run a macro and ask it to
-  // produce augmentations.
+  /// Handle requests that are for the host.
+  @override
+  Future<Object?> handle(Object request) async {
+    // TODO(davidmorgan): don't assume the type. Return `null` for types
+    // that should be passed through to the service that was passed in.
+    final macroStartedRequest =
+        MacroStartedRequest.fromJson(request as Map<String, Object?>);
+    _macroPhases = macroStartedRequest.macroDescription.runsInPhases.toSet();
```

Comment:
Yeah I think just completing a completer here would be better for sure (and fix the issue above)